### ISSUE TITLE
Why are there no proper epochs?

### ISSUE THREAD

**Issue by SirRob1997** [facebookresearch/DomainBed#17]
As of now, the training loop just uses continuous update steps and hence the checkpoints usually occur in the middle of epochs as reported by the floats produced in `'epoch': step / steps_per_epoch` .

For me, this doesn't make too much sense. Shouldn't we count proper epochs and then define `checkpoint_freq` in terms of the epochs instead of the steps such that we report validation accuracies for every `checkpoint_freq` COMPLETED epochs?

This way we could also follow proper design for learning rate schedules since `scheduler.step(val_loss)` is usually invoked after every epoch and for `torch.optim.lr_scheduler.ReduceLROnPlateau` the learning rate gets adapted based on the change of the validation loss over multiple consecutive epochs.

**Comment by igul222** [comment]
This seems closed now, but for reference doing a proper epoch loop breaks IRM since it relies on independent examples in each minibatch and sampling without replacement introduces a slight correlation between examples. The way this manifests is that if you train for long enough, the penalty term becomes negative. I think you should be able to invoke `scheduler.step(val_loss)` every checkpoint_freq steps.

Separately, we found it more practical to define `checkpoint_freq` and other such things in terms of steps rather than epochs because we're dealing with many datasets of very different sizes. As one example, training for 5000 steps is a reasonable default, whereas training for 30 epochs would simultaneously be too short for MNIST and too long for DomainNet.

**Comment by SirRob1997** [comment]
Yeah, came to the same conclusion. Closed due to the mentioned reasons.
